cairo/crates/token/src/extensions/hyp_erc20_dex_collateral.cairo

```diff
@@ -0,0 +1,239 @@
+    #[abi(embed_v0)]
+    impl HypErc20CollateralImpl =
+        HypErc20CollateralComponent::HypErc20CollateralImpl<ContractState>;
+    impl HypErc20CollateralInternalImpl =
+        HypErc20CollateralComponent::HypErc20CollateralInternalImpl<ContractState>;
+    // Upgradeable
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+    // TokenRouter
+    #[abi(embed_v0)]
+    impl TokenRouterImpl = TokenRouterComponent::TokenRouterImpl<ContractState>;
+
+    #[storage]
+    struct Storage {
+        dex: ContractAddress,
+        #[substorage(v0)]
+        collateral: HypErc20CollateralComponent::Storage,
+        #[substorage(v0)]
+        mailbox: MailboxclientComponent::Storage,
+        #[substorage(v0)]
+        token_router: TokenRouterComponent::Storage,
+        #[substorage(v0)]
+        gas_router: GasRouterComponent::Storage,
+        #[substorage(v0)]
+        router: RouterComponent::Storage,
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+        #[substorage(v0)]
+        upgradeable: UpgradeableComponent::Storage,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        HypErc20CollateralEvent: HypErc20CollateralComponent::Event,
+        #[flat]
+        MailBoxClientEvent: MailboxclientComponent::Event,
+        #[flat]
+        GasRouterEvent: GasRouterComponent::Event,
+        #[flat]
+        RouterEvent: RouterComponent::Event,
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+        #[flat]
+        TokenRouterEvent: TokenRouterComponent::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event,
+        DexDeposit: DexDeposit,
+    }
+
+    // An event that is emitted when the DEX deposit is successful (via transfer_to_hook)
+    #[derive(Drop, starknet::Event)]
+    struct DexDeposit {
+        #[key]
+        token: ContractAddress,
+        #[key]
+        recipient: ContractAddress,
+        amount: u256,
+    }
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        mailbox: ContractAddress,
+        dex: ContractAddress,
+        wrapped_token: ContractAddress,
+        owner: ContractAddress,
+        hook: ContractAddress,
+        interchain_security_module: ContractAddress,
+    ) {
+        self.ownable.initializer(owner);
+        self
+            .mailbox
+            .initialize(mailbox, Option::Some(hook), Option::Some(interchain_security_module));
+        self.dex.write(dex);
+        self.collateral.initialize(wrapped_token);
+    }
+
+    impl TokenRouterHooksTraitImpl of TokenRouterHooksTrait<ContractState> {
+        fn transfer_from_sender_hook(
+            ref self: TokenRouterComponent::ComponentState<ContractState>, amount_or_id: u256,
+        ) -> Bytes {
+            HypErc20CollateralComponent::TokenRouterHooksImpl::transfer_from_sender_hook(
+                ref self, amount_or_id,
+            )
+        }
```

Comment:
is the sender expected to withdraw from the exchange before using the warp route in this direction?

Reply:
Yes, that's correct.

Reply:
iiuc that will require 3 different transactions by the user
1. withdraw
2. approve
3. transfer

Reply:
Yes, that's correct.